packages/stellar-axelar-std-derive/src/storage.rs

```diff
@@ -9,9 +12,122 @@ enum Value {
     Type(Type),
 }
 
-struct StorageAttributes {
-    storage_type: StorageType,
-    value: Value,
+impl TryFrom<&[Attribute]> for Value {
+    type Error = String;
+
+    fn try_from(attrs: &[Attribute]) -> Result<Self, Self::Error> {
+        let attr = attrs
+            .iter()
+            .filter(|attr| attr.path().is_ident("status") || attr.path().is_ident("value"))
+            .exactly_one()
+            .map_err(|_| {
+                "exactly one of #[status] and #[value(Type)] must be provided".to_string()
+            })?;
+
+        if let Meta::List(list) = &attr.meta {
+            if attr.path().is_ident("value") {
+                Ok(Self::Type(
+                    list.parse_args::<Type>()
+                        .map_err(|_| "failed to parse value type".to_string())?,
+                ))
+            } else {
+                Err("status attribute cannot have parameters".to_string())
+            }
+        } else if attr.path().is_ident("status") {
+            Ok(Self::Status)
+        } else {
+            Err("value attribute must contain a type parameter: #[value(Type)]".to_string())
+        }
+    }
+}
+
+trait VariantExt {
+    fn storage_params(&self) -> TokenStream;
+    fn storage_key(&self, enum_name: &Ident) -> TokenStream;
+}
+
+impl VariantExt for Variant {
+    /// Returns the parameter list for a storage enum variant.
+    fn storage_params(&self) -> TokenStream {
+        let (field_names, field_types) = fields_data(&self.fields);
+
+        if field_names.is_empty() {
+            quote! { env: &soroban_sdk::Env }
+        } else {
+            quote! { env: &soroban_sdk::Env, #(#field_names: #field_types),* }
+        }
+    }
+
+    /// Returns the key for a storage enum variant.
+    fn storage_key(&self, enum_name: &Ident) -> TokenStream {
+        let (field_names, _) = fields_data(&self.fields);
```

Comment:
```suggestion
        let field_names = self.fields_data();
```
should be a trait method as well. And do you need the other return value?